src/query/storage/completed_tags.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import (
+	"bytes"
+	"errors"
+	"sort"
+)
+
+// MergeWith merges two CompleteTagsResults together, handling deduplication.
+// Once all merges are completed, Finalize() should be called to give sorted results.
+func (r *CompleteTagsResult) MergeWith(other *CompleteTagsResult) error {
```

Comment:
this is a bit weird design to me. Consider having a CompleteTagsBuilder which has this method. Instead of Finalize, the building can have Build() method which returns the CompleteTagsResult. This way you don't end up having things like seenMap defined on the CompleteTagsResult struct

Reply:
Refactored it to use a builder. I think the builder approach is likely a little less efficient, but much cleaner to read through and use

Reply:
less efficient how ?

Reply:
This approach is creating additional data structures to do the value merge rather than doing it on the tags themselves

Pretty minimal impact overall and this isn't a hot path so not worried